h2o-core/src/test/java/water/fvec/VecTest.java

```diff
@@ -148,4 +146,50 @@ public void map(Chunk c) {
     source.remove(new Futures()).blockForPending();
     con.remove(new Futures()).blockForPending();
   }
+
+  @Test public void testChunkForChunkIdxAfterVecUpdate() {
+    try {
+      Scope.enter();
+      Vec v1 = Scope.track(makeSeq(2 * FileVec.DFLT_CHUNK_SIZE, false));
+      Vec v2 = Scope.track(makeSeq(2 * FileVec.DFLT_CHUNK_SIZE, false));
+      Chunk c0 = v1.chunkForChunkIdx(0);
+      c0._vec = v2; // inject any vec into the cached POJO, this simulates a stale Vec reference
+      Chunk c0c = v1.chunkForChunkIdx(0);
+      assertSame(c0, c0c);
+      // the vec reference was updated in the existing live object
+      assertSame(v1, c0c._vec);
+    } finally {
+      Scope.exit();
+    }
+  }
+
+  @Test public void testChunkForChunkIdxMRTask() {
+    Assume.assumeTrue(H2O.getCloudSize() > 0);
+    try {
+      Scope.enter();
+      Vec v = Scope.track(makeConN((long) 1e6, 16));
+      // 1. Run an arbitrary MRTask to populate the POJO Chunk caches
+      new MRTask() {
+        @Override
+        public void map(Chunk c) {
+          if (c.vec().get_type() != Vec.T_NUM)
+            throw new IllegalStateException("Expected a numeric Vec");
+        }
+      }.doAll(v);
+      // 2. Install updated Vec in DKV
+      v._type = Vec.T_TIME;
+      DKV.put(v);
+      // 3. Run another MRTask, we should get updated Vec on all nodes not the stale one
+      new MRTask() {
+        @Override
+        public void map(Chunk c) {
+          if (c.vec().get_type() != Vec.T_TIME)
+            throw new IllegalStateException("Expected a time Vec");
+        }
+      }.doAll(v);
```

Comment:
@michalkurka Where will this `v` is going to come from when the MRTask is deserialised on the remote node for execution?

Reply:
When MRTask is submitted, only keys are actually sent to the remote nodes. The values of keys (Vecs) are then fetched if they are not already present on the remote.

`DKV.put` on line 181 invalidates local replicas for the `v._key` - this means some nodes will have to get the Vec from a remote. Either way, the MRTask and the nodes will have a correct (=cluster consistent) instance of the Vec. However, none of the invalidation is applied to POJOs of Vecs that referenced in Chunks.

Reply:
@michalkurka thanks, it looks like my understanding described  in https://github.com/h2oai/h2o-3/pull/3480/files#r288684152 comment is correct.

Reply:
Yep, it is correct :)